commit 294f6cf48666825d23c9372ef37631232746e40d
Author: Timo Warns <Warns@pre-sense.de>
Date:   Fri Feb 25 14:44:21 2011 -0800

    ldm: corrupted partition table can cause kernel oops
    
    The kernel automatically evaluates partition tables of storage devices.
    The code for evaluating LDM partitions (in fs/partitions/ldm.c) contains
    a bug that causes a kernel oops on certain corrupted LDM partitions.  A
    kernel subsystem seems to crash, because, after the oops, the kernel no
    longer recognizes newly connected storage devices.
    
    The patch changes ldm_parse_vmdb() to Validate the value of vblk_size.
    
    Signed-off-by: Timo Warns <warns@pre-sense.de>
    Cc: Eugene Teo <eugeneteo@kernel.sg>
    Acked-by: Richard Russon <ldm@flatcap.org>
    Cc: Harvey Harrison <harvey.harrison@gmail.com>
    Cc: <stable@kernel.org>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
    [dannf: backported to Debian's 2.6.26]

diff --git a/fs/partitions/ldm.c b/fs/partitions/ldm.c
index 0fdda2e..2ebdde7 100644
--- a/fs/partitions/ldm.c
+++ b/fs/partitions/ldm.c
@@ -251,6 +251,11 @@ static bool ldm_parse_vmdb (const u8 *data, struct vmdb *vm)
 	}
 
 	vm->vblk_size     = BE32 (data + 0x08);
+	if (vm->vblk_size == 0) {
+		ldm_error ("Illegal VBLK size");
+		return false;
+	}
+
 	vm->vblk_offset   = BE32 (data + 0x0C);
 	vm->last_vblk_seq = BE32 (data + 0x04);
 
